File: csharp/src/VendorConnector/Services/FirmwarePolicy.cs
using System;

namespace VendorConnector.Services
{
    public static class FirmwarePolicy
    {
        public static bool CanUpdate(string nowLocal, string window)
        {
            var parts = window.Split('-');
            var from = TimeSpan.Parse(parts[0]);
            var to   = TimeSpan.Parse(parts[1]);
            var now  = TimeSpan.Parse(nowLocal);
            if (from <= to) return now >= from && now <= to;
            return now >= from || now <= to;
        }

        public static int SemverCmp(string a, string b)
        {
            var pa = a.Split('.');
            var pb = b.Split('.');
            for (int i = 0; i < 3; i++)
            {
                int ai = int.Parse(pa[i]);
                int bi = int.Parse(pb[i]);
                if (ai != bi) return ai.CompareTo(bi);
            }
            return 0;
        }
    }
}

File: js/vendor-connector/src/firmwarePolicy.js
export function canUpdate(nowLocal, window){
  const [from, to] = window.split("-");
  const [nh, nm] = nowLocal.split(":").map(Number);
  const [fh, fm] = from.split(":").map(Number);
  const [th, tm] = to.split(":").map(Number);
  const now = nh*60+nm, f = fh*60+fm, t = th*60+tm;
  if (f <= t) return now >= f && now <= t;
  return now >= f || now <= t;
}
export function semverCmp(a,b){
  const pa=a.split('.').map(Number), pb=b.split('.').map(Number);
  for (let i=0;i<3;i++){ if (pa[i]!==pb[i]) return pa[i]-pb[i]; }
  return 0;
}
